contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -344,13 +345,17 @@ bool Filter::doTrailer(ProcessorState& state, Http::HeaderMap& trailers) {
     // do data first
     if (!state.isBufferDataEmpty()) {
       done = doDataGo(state, state.getBufferData(), false);
-    }
-    // NP: can not use done as condition here, since done will be false
-    // maybe we can remove the done variable totally? by using state_ only?
-    // continue trailers
-    if (state.state() == FilterState::WaitingTrailer) {
+      // NP: can not use done as condition here, since done will be false
+      // maybe we can remove the done variable totally? by using state_ only?
+      // continue trailers
+      if (state.state() == FilterState::WaitingTrailer) {
+        done = doTrailerGo(state, trailers);
+      }
+    } else {
       done = doTrailerGo(state, trailers);
     }
+    state.doDataList.moveOut(body);
```

Comment:
emm, line after 347 is not  
```
if (done) {
      state.continueDoData();
}
```
https://github.com/envoyproxy/envoy/blob/0a2cf15c179379c877ed0825be6c8849b0d9bd20/contrib/golang/filters/http/source/golang_filter.cc#LL347C6-L347C6

Reply:
yep, there is not yet, I mean we should add it there, and we do not these two new lines then.

Reply:
emmm, i think `moveOut` should not be there, since even if `isBufferDataEmpty` is true, `doDataList` may be also not empty, for example, `decodeHeaders` returns `continue` and `decodeData` returns `StopAndBuffer`, `doDataGo` will be called at `decodeData` phase first, so `doDataList` will not be empty at `decodeTrailers` phase. Did i make myself clear? 😄

Reply:
oh, yes, the case you said is right.
but, the current implemention is not a proper fix. for example, `doDataGo` may return `false` at line 347, and we should not send doDataList then.
I think the proper fix might be, add `continueDoData` in the else branch, before `doTrailerGo` at line 355.

Reply:
Also, could you please add some related test case? thanks very much.
Ah, seem this PR is bugfix more than feature, cool!